text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+
+The basic tension is this: the ergonomics and safety advantages that Rust offers
+work by explicitly preventing certain behaviour, such as the aliasing of mutable
+variables. We cannot translate idiomatic C++ directly into Rust and get safety;
+we get safety by ruling out certain patterns.
+
+I think it is really worth stepping back and considering our goals. We
+would like to have a tool for authoring & manipulating font files, and we would
+like it to be a) fast and b) easy to use. This is something that we can achieve,
+but I do not think the result is going to look like it might look in C++. My
+feeling is that too much of our thinking (or at least my thinking!) so far has
+been focused on porting an existing design to Rust, and that this may be
+misguided.
+
+#### Subsetting is special
+
+A major focus of (and motivation for) this work has been hb-subset. I think this
+has influenced our thinking in important ways. I'm not entirely confident in my
+analysis here, so please correct any bad assumptions. With that said:
+
+- My sense is that the design of hb-subset is largely path-dependent on the
+  existing design of HarfBuzz. This is just an intuition. What this means to me,
+  though, is that if we were designing a subsetter from scratch, we might make
+  different choices.
+- Subsetting really feels like a special case. To the best of my understanding
+  it is used almost exclusively(?) by Google Fonts, and at a much larger scale.
+- I think that trying to address the *general* case in a way that also natively
+  supports the subsetting case might be a mistake. In particular I worry that
+  this dramatically constrains our solution space, and will lead to a design
+  that is seriously compromised in the general case.
+- I would prefer to focus on the general case, and then try to figure out a more
+  holistic approach to the subsetting problem. In practical terms, this means a
+  general approach that does more transient allocation for subtables etc, and
+  then a subsetting approach that (for instance) uses a long-running process to
+  handle multiple subsetting requests, and reuses memory and other state between
+  them.
+
+## next steps
+
+### a design for mutability
+
+My main concern right now is figuring out a model for compilation & mutability.
+For compilation, I'm currently imagining generating builder types, but I haven't
+dug into this yet.
+
+Offset resolution and rewriting will also be an important consideration. A
+compiler author should be able to use these tools to do things like deduplicate
+identical subtables, or to reorder tables as needed, which means we do need a
+mechanism for rewriting offsets as needed. There are various options here; for
+instance we could track offsets by ids, and then keep track of the location of
+the offsets as we write tables into a shared buffer, rewriting them later as
+needed. This will need more thought, experimentation, and discussion.
```

Comment:
@garretrieger is probably the authority here (https://github.com/harfbuzz/harfbuzz/blob/main/docs/repacker.md) and can advise us fairly specifically on how to tackle this.

Reply:
Here's a link to the presentation I mentioned: https://docs.google.com/presentation/d/1_AUjPtay6B41e-QHtLjxmf9VpgghHXjpNMFJtq1Fy08/edit?usp=sharing